Python/pystate.c

```diff
@@ -1057,6 +1057,47 @@ _PyErr_SetInterpreterAlreadyRunning(void)
     PyErr_SetString(PyExc_InterpreterError, "interpreter already running");
 }
 
+int
+_PyInterpreterState_IsShuttingDown(PyInterpreterState *interp)
+{
+    return get_main_thread(interp) == _PyInterpreterState_MAIN_SHUTTING_DOWN;
+}
+
+int
+_PyInterpreterState_SetShuttingDown(PyInterpreterState *interp)
+{
+    void *expected = NULL;
+    // XXX Use an exchange without a comparison?
+    if (_Py_atomic_compare_exchange_ptr(&interp->threads.main,
+                                        &expected,
+                                        _PyInterpreterState_MAIN_SHUTTING_DOWN) == 0)
+    {
+        /* We're either already shutting down or another thread started running */
+        _PyErr_SetInterpreterAlreadyRunning();
+        return -1;
+    }
+    assert(_PyInterpreterState_IsShuttingDown(interp));
+    assert(!_PyInterpreterState_IsRunningMain(interp));
+    /* At this point, we're certain that no other threads can set the main thread.
+     * However, there might be some remaining thread states--in that case, let's just raise.
+     * We could wait for them all to finish up, but that's a job for later. */
+
+    // TODO: Switch this to the per-interpreter lock once Eric's PR is merged
+    HEAD_LOCK(interp->runtime);
+    PyThreadState *thread_head = interp->threads.head;
+    HEAD_UNLOCK(interp->runtime);
+    if (interp->threads.head != NULL)
```

Comment:
this

Reply:
Good catch! I thought I fixed that when I re-added my changes, I guess not. Thank you :)